File: sdks/typescript/src/crypto/signerHelpers.ts
/**
 * Convenience wrappers to create XacheSigner from common wallet types.
 *
 * These helpers make it easy to use external wallets (ethers.js, Solana Keypair,
 * Coinbase AgentKit) as XacheSigner instances without manual interface mapping.
 *
 * @example
 * ```typescript
 * import { createSignerFromEthersWallet, XacheClient } from '@xache/sdk';
 * import { ethers } from 'ethers';
 *
 * const wallet = new ethers.Wallet(privateKey);
 * const signer = createSignerFromEthersWallet(wallet);
 * const client = new XacheClient({ apiUrl, did, signer });
 * ```
 */

import type { XacheSigner } from '../types';

/**
 * Create XacheSigner from an ethers.js v6 Wallet.
 *
 * @param wallet - ethers.js Wallet instance (must have address, signTypedData, signingKey)
 */
export function createSignerFromEthersWallet(wallet: {
  address: string;
  signTypedData: (
    domain: Record<string, unknown>,
    types: Record<string, Array<{ name: string; type: string }>>,
    value: Record<string, unknown>
  ) => Promise<string>;
  signingKey: {
    sign: (digest: string) => { serialized: string };
  };
}): XacheSigner {
  return {
    getAddress: async () => wallet.address,
    signTypedData: (domain, types, value) =>
      wallet.signTypedData(domain as Record<string, unknown>, types, value),
    signMessage: async (message: Uint8Array) => {
      // For auth headers: message is a raw keccak256 hash (32 bytes for EVM)
      // or raw message bytes (for Solana, though this wrapper is EVM-specific)
      // Use signingKey.sign for raw hash signing (no EIP-191 prefix)
      const hashHex = '0x' + Buffer.from(message).toString('hex');
      const sig = wallet.signingKey.sign(hashHex);
      const sigHex = sig.serialized.startsWith('0x') ? sig.serialized.slice(2) : sig.serialized;
      return new Uint8Array(Buffer.from(sigHex, 'hex'));
    },
  };
}

/**
 * Create XacheSigner from a Solana Keypair.
 *
 * @param keypair - Solana Keypair instance (from @solana/web3.js)
 */
export function createSignerFromSolanaKeypair(keypair: {
  publicKey: { toBase58(): string };
  secretKey: Uint8Array;
}): XacheSigner {
  return {
    getAddress: async () => keypair.publicKey.toBase58(),
    signTypedData: async () => {
      throw new Error('EVM signTypedData is not supported on Solana Keypair');
    },
    signMessage: async (message: Uint8Array) => {
      // Use tweetnacl for ed25519 detached signing
      const nacl = await import('tweetnacl');
      return nacl.default.sign.detached(message, keypair.secretKey);
    },
  };
}

/**
 * Create XacheSigner from a Coinbase AgentKit instance.
 *
 * @param agentKit - AgentKit instance with getWallet() method
 */
export function createSignerFromAgentKit(agentKit: {
  getWallet(): Promise<{
    address: string;
    signTypedData(
      domain: Record<string, unknown>,
      types: Record<string, Array<{ name: string; type: string }>>,
      value: Record<string, unknown>
    ): Promise<string>;
    signMessage(message: string | Uint8Array): Promise<string>;
  }>;
}): XacheSigner {
  let wallet: Awaited<ReturnType<typeof agentKit.getWallet>> | undefined;

  const getWallet = async () => {
    if (!wallet) wallet = await agentKit.getWallet();
    return wallet;
  };

  return {
    getAddress: async () => (await getWallet()).address,
    signTypedData: async (domain, types, value) =>
      (await getWallet()).signTypedData(domain as Record<string, unknown>, types, value),
    signMessage: async (message: Uint8Array) => {
      const w = await getWallet();
      const sig = await w.signMessage(message);
      const sigHex = sig.startsWith('0x') ? sig.slice(2) : sig;
      return new Uint8Array(Buffer.from(sigHex, 'hex'));
    },
  };
}

File: sdks/python/xache/crypto/signer_helpers.py
"""
Convenience wrappers for creating XacheSigner from common wallet types.

Example:
    from xache import XacheClient, create_signer_from_eth_account
    from eth_account import Account

    acct = Account.from_key(private_key)
    signer = create_signer_from_eth_account(acct)
    client = XacheClient(api_url=..., did=..., signer=signer)
"""

from __future__ import annotations

from typing import Any


def create_signer_from_eth_account(account: Any) -> Any:
    """
    Create XacheSigner from an eth_account.Account instance.

    Args:
        account: eth_account.Account instance (from Account.from_key())

    Returns:
        XacheSigner-compatible object for use with XacheClient(signer=...)

    Example:
        from eth_account import Account
        acct = Account.from_key("0x...")
        signer = create_signer_from_eth_account(acct)
        client = XacheClient(api_url=..., did=..., signer=signer)
    """

    class EthAccountSigner:
        async def get_address(self) -> str:
            return account.address

        async def sign_message(self, message: bytes) -> bytes:
            from eth_keys import keys as eth_keys  # type: ignore[attr-defined]

            pk = eth_keys.PrivateKey(account.key)
            # Sign the raw hash (message is already keccak256 hash for EVM auth)
            signature = pk.sign_msg_hash(message)
            r = signature.r.to_bytes(32, "big")
            s = signature.s.to_bytes(32, "big")
            v = bytes([signature.v + 27])
            return r + s + v

        async def sign_typed_data(self, domain, types, value) -> str:
            # Use eth_account's built-in EIP-712 signing
            from eth_account.messages import encode_typed_data

            # Build the full typed data structure
            primary_type = next(iter(types.keys()))
            full_message = {
                "domain": domain,
                "types": types,
                "primaryType": primary_type,
                "message": value,
            }
            signed = account.sign_message(encode_typed_data(full_message=full_message))
            return signed.signature.hex()

    return EthAccountSigner()


def create_signer_from_solana_keypair(keypair: Any) -> Any:
    """
    Create XacheSigner from a solders.keypair.Keypair instance.

    Args:
        keypair: solders.keypair.Keypair instance

    Returns:
        XacheSigner-compatible object for use with XacheClient(signer=...)

    Example:
        from solders.keypair import Keypair
        kp = Keypair.from_seed(seed_bytes)
        signer = create_signer_from_solana_keypair(kp)
        client = XacheClient(api_url=..., did=..., signer=signer)
    """

    class SolanaKeypairSigner:
        async def get_address(self) -> str:
            return str(keypair.pubkey())

        async def sign_message(self, message: bytes) -> bytes:
            sig = keypair.sign_message(message)
            return bytes(sig)

        async def sign_typed_data(self, domain, types, value) -> str:
            raise NotImplementedError(
                "EVM signTypedData is not supported on Solana Keypair"
            )

    return SolanaKeypairSigner()
